config/management/src/verify.rs

```diff
@@ -110,7 +146,134 @@ fn write_waypoint(storage: &dyn Storage, buffer: &mut String, key: &str) {
                     .unwrap_or_else(|_| "Invalid waypoint".into())
             }
         })
-        .unwrap_or_else(|e| format!("{:?}", e));
+        .unwrap_or_else(|e| e.to_string());
 
     writeln!(buffer, "{} - {}", key, value).unwrap();
 }
+
+fn compare_genesis(
+    storage: &dyn Storage,
+    buffer: &mut String,
+    genesis_path: &PathBuf,
+) -> Result<(), Error> {
+    let db_path = TempPath::new();
+    let (db_rw, expected_waypoint) = compute_genesis(genesis_path, db_path.path())?;
+    let validator_account = validator_account(storage)?;
+    let validator_config = validator_config(validator_account, db_rw.reader.clone())?;
+
+    let actual_consensus_key = ed25519_from_storage(CONSENSUS_KEY, storage)
+        .map_err(|e| Error::LocalStorageReadError(CONSENSUS_KEY, e))?;
+    let expected_consensus_key = validator_config.consensus_public_key;
+    write_assert(
+        buffer,
+        CONSENSUS_KEY,
+        actual_consensus_key == expected_consensus_key,
+    );
+
+    let actual_validator_key = x25519_from_storage(VALIDATOR_NETWORK_KEY, storage)
+        .map_err(|e| Error::LocalStorageReadError(VALIDATOR_NETWORK_KEY, e))?;
+    let expected_validator_key = validator_config.validator_network_identity_public_key;
+    write_assert(
+        buffer,
+        VALIDATOR_NETWORK_KEY,
+        actual_validator_key == expected_validator_key,
+    );
+
+    let actual_fullnode_key = x25519_from_storage(FULLNODE_NETWORK_KEY, storage)
+        .map_err(|e| Error::LocalStorageReadError(VALIDATOR_NETWORK_KEY, e))?;
+    let expected_fullnode_key = validator_config.full_node_network_identity_public_key;
+    write_assert(
+        buffer,
+        FULLNODE_NETWORK_KEY,
+        actual_fullnode_key == expected_fullnode_key,
+    );
+
+    let actual_waypoint = storage
```

Comment:
nit: I'd be tempted to compare the waypoint first in this method as it means that "expected_waypoint" is used closer to where. This comparison also doesn't rely on validator_config

Reply:
Interesting point....